File: 112-path-sum/112.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:
        if not root:
            return False

        def dfs(path, root):
            if not root.left and not root.right:
                return sum(path + [root.val]) == targetSum

            if root.left and root.right:
                return dfs(path + [root.val], root.left) or dfs(path + [root.val], root.right)
            elif root.left:
                return dfs(path + [root.val], root.left)
            elif root.right:
                return dfs(path + [root.val], root.right)

        return dfs([], root)

File: 112-path-sum/112.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
       if (root == null) {
           return false;
       }

       return dfs(0, root, targetSum);
    }
    
    private boolean dfs(int cur_sum, TreeNode root, int targetSum) {
        if (root.left == null && root.right == null) {
            return cur_sum + root.val == targetSum;
        }

        if (root.left != null && root.right != null) {
            return dfs(cur_sum + root.val, root.left, targetSum) || dfs(cur_sum + root.val, root.right, targetSum);
        } else if(root.left != null) {
            return dfs(cur_sum + root.val, root.left, targetSum);
        } else if (root.right != null) {
            return dfs(cur_sum + root.val, root.right, targetSum);
        }

        return false;
    }
}
